src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt

```diff
@@ -0,0 +1,116 @@
+            "application/json",
+        ).tuner {
+            it.setRequestProperty("Authorization", "Bearer $token")
+        }
+
+        // Add system prompt
+        if (chatHistory.size == 1) {
+            chatHistory[0] = ChatUserMessage(
+                createInstructionPrompt(
+                    chatHistory[0].content,
+                ),
+            )
+        }
+
+        val llmRequestBody = HuggingFaceRequestBody(chatHistory, Parameters(topProbability, temperature)).toMap()
+        var sendResult = SendResult.OK
+        try {
+            httpRequest.connect {
+                it.write(GsonBuilder().disableHtmlEscaping().create().toJson(llmRequestBody))
+                when (val responseCode = (it.connection as HttpURLConnection).responseCode) {
+                    HttpURLConnection.HTTP_OK -> {
+                        val text = it.reader.readLine()
+                        val generatedTestCases = extractLLMGeneratedCode(
+                            JsonParser.parseString(text).asJsonArray[0]
+                                .asJsonObject["generated_text"].asString.trim(),
+                        )
+                        (testsAssembler as JUnitTestsAssembler).consume(generatedTestCases)
+                    }
+
+                    HttpURLConnection.HTTP_INTERNAL_ERROR -> {
+                        llmErrorManager.errorProcess(
+                            LLMMessagesBundle.get("serverProblems"),
+                            project,
+                            errorMonitor,
+                        )
+                        sendResult = SendResult.OTHER
+                    }
+
+                    HttpURLConnection.HTTP_BAD_REQUEST -> {
+                        llmErrorManager.errorProcess(
+                            LLMMessagesBundle.get("hfServerError"),
+                            project,
+                            errorMonitor,
+                        )
+                        sendResult = SendResult.OTHER
+                    }
+                }
+            }
+        } catch (e: HttpStatusException) {
+            log.error { "Error in sending request: ${e.message}" }
+        }
+        return sendResult
+    }
+
+    /**
+     * Creates the required prompt for Llama models. For more details see:
+     * https://huggingface.co/blog/llama2#how-to-prompt-llama-2
+     */
+    private fun createInstructionPrompt(userMessage: String): String {
+        // TODO: This is Llama-specific and should support other LLMs hosted on HF too.
+        return "<s>[INST] <<SYS>> ${LLMDefaultsBundle.get("huggingFaceInitialSystemPrompt")} <</SYS>> $userMessage [/INST]"
+    }
+
+    /**
+     * Extracts code blocks in LLMs' response.
+     * Also, it handles the cases where the LLM-generated code does not end with ```
+     */
+    private fun extractLLMGeneratedCode(text: String): String {
+        val modifiedText = text.replace("```java", "```").replace("````", "```")
```

Comment:
Also, one point: the `java` is hard-coded here. We need to figure out the language in which the test generation was requested dynamically. But let's have another PR for this, as it is not that trivial and may require intrusive changes; a `TODO` will be enough for now here.

Reply:
Why is `.replace("````", "```")` needed? It seems not to affect anything. 

I've never seen the LLM responding with four backtick blocks, although it may merge several triple backtick blocks, resulting in the `"``````"` block.

Reply:
I've seen it twice with  ` ```` `. It can be removed, though.